### PR TITLE
Define 'pinch' as a counterpart to 'punchIn'

### PR DESCRIPTION
**Pull request by TOTBWF**
## Patch Description

During the process of implementing https://github.com/agda/agda-categories/pull/274, @JacquesCarette pointed out that one of the functions I implemented existed in the stdlib under the name `punchIn`. However, this injective function has a surjective counterpart that seems to be missing. This PR adds this function under the name `pinch`, and proves some simple properties about it.

## Notes
I'm not 100% attached to the name, perhaps `pinchIn` would be more appropriate to highlight the connection to `punchIn`. There are also a handful of identities that we can prove relating `pinch` to `punchIn` (See https://en.wikipedia.org/wiki/Simplicial_set#Face_and_degeneracy_maps and https://github.com/agda/agda-categories/pull/274/files#diff-d758c95623d24b748cbba2280dcde56f96e72c64c271bab9afdc4e068efe8117R114), though these seem a bit domain specific for the stdlib.

Furthermore, any map `Fin m -> Fin n` can be factored into a canonical form of a bunch of calls to `pinch` followed a bunch of calls to `punchIn`, though after some preliminary attempts this is proving to be Very Difficult. Perhaps someone braver than I could give it a go.

### CODE REVIEW

**Comment by JacquesCarette** [comment]
Small comment: any *order preserving* map `Fin m -> Fin n` can be so factored. Otherwise, all non-identity permutations are counter-examples...

**Comment by TOTBWF** [comment]
Doh! Been thinking too much about simplices, I forgot that non-monotonic maps even existed for a second!

File: src/Data/Fin/Base.agda [hunk 1]
```diff
@@ -240,6 +240,14 @@ punchIn zero    j       = suc j
 punchIn (suc i) zero    = zero
 punchIn (suc i) (suc j) = suc (punchIn i j)
 
+-- The surjection f(i,j) such that f(i,i) = f(i,i+1) = i
+
+pinch : ∀ {m} → Fin m → Fin (suc m) → Fin m
+pinch zero zero = zero
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Base.agda:246]: Should probably collapse 1st and 3rd cases into `pinch _ zero = zero`?

File: src/Data/Fin/Base.agda [hunk 2]
```diff
@@ -240,6 +240,14 @@ punchIn zero    j       = suc j
 punchIn (suc i) zero    = zero
 punchIn (suc i) (suc j) = suc (punchIn i j)
 
+-- The surjection f(i,j) such that f(i,i) = f(i,i+1) = i
+
+pinch : ∀ {m} → Fin m → Fin (suc m) → Fin m
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Base.agda:245]: We usually use `n` for natural numbers when we're only using one of them. I realise we're not that consistent in this particular file!

File: src/Data/Fin/Base.agda [hunk 3]
```diff
@@ -240,6 +240,14 @@ punchIn zero    j       = suc j
 punchIn (suc i) zero    = zero
 punchIn (suc i) (suc j) = suc (punchIn i j)
 
+-- The surjection f(i,j) such that f(i,i) = f(i,i+1) = i
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Base.agda:243]: This comment doesn't explain the behaviour of the function when j != i or i+1?

File: src/Data/Fin/Properties.agda [hunk 4]
```diff
@@ -675,6 +675,23 @@ punchOut-punchIn (suc i) {suc j} = cong suc (begin
   j                                                           ∎)
   where open ≡-Reasoning
 
+
+------------------------------------------------------------------------
+-- pinch
+------------------------------------------------------------------------
+
+pinch-surjective : ∀ {m} i (j : Fin m) → Σ[ k ∈ Fin (suc m) ] (pinch i k ≡ j)
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Properties.agda:683]: Can we use the actual `Surjective` type from `Function`? Helps ensure consistency and therefore compatibility between different parts of the library!

File: src/Data/Fin/Properties.agda [hunk 5]
```diff
@@ -675,6 +675,23 @@ punchOut-punchIn (suc i) {suc j} = cong suc (begin
...
+pinch-surjective 0F 0F = zero , refl
+pinch-surjective 0F (suc j) = (suc (suc j)) , refl
+pinch-surjective (suc i) 0F = zero , refl
+pinch-surjective (suc i) (suc j) = map suc (cong suc) (pinch-surjective i j)
+
+pinch-mono-≤ : ∀ {m} (i : Fin m) → (pinch i) Preserves _≤_ ⟶ _≤_
+pinch-mono-≤ 0F {0F} {k} 0≤n = z≤n
+pinch-mono-≤ 0F {suc j} {suc k} (s≤s j≤k) = j≤k
+pinch-mono-≤ (suc i) {0F} {k} 0≤n = z≤n
+pinch-mono-≤ (suc i) {suc j} {suc k} (s≤s j≤k) = s≤s (pinch-mono-≤ i j≤k)
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Properties.agda:693]: I don't think you need to pattern match on the `j` and `k` implicits in this proof?

File: src/Data/Fin/Properties.agda [hunk 6]
```diff
@@ -675,6 +675,23 @@ punchOut-punchIn (suc i) {suc j} = cong suc (begin
...
   where open ≡-Reasoning
 
+
+------------------------------------------------------------------------
+-- pinch
+------------------------------------------------------------------------
+
+pinch-surjective : ∀ {m} i (j : Fin m) → Σ[ k ∈ Fin (suc m) ] (pinch i k ≡ j)
+pinch-surjective 0F 0F = zero , refl
+pinch-surjective 0F (suc j) = (suc (suc j)) , refl
```
Comment (MatthewDaggitt) [inline src/Data/Fin/Properties.agda:685]: No need for the outer brackets round `suc (suc j)`

**Review by MatthewDaggitt — COMMENTED** [review]
> I'm not 100% attached to the name, perhaps pinchIn would be more appropriate to highlight the connection to punchIn

Eh, I don't really like the names `punchIn` or `punchOut` either. I think `pinch` is fine for the moment, unless you think there's an analogous `pinch` function for `punchOut` as well?

> There are also a handful of identities that we can prove relating pinch to punchIn ... though these seem a bit domain specific for the stdlib.

Anything that directly relates the two functions is fair game for inclusion into the library. If you think some of them would be useful then feel to add them to the PR.

> Furthermore, any map Fin m -> Fin n can be factored into a canonical form of a bunch of calls to pinch followed a bunch of calls to punchIn, though after some preliminary attempts this is proving to be Very Difficult. Perhaps someone braver than I could give it a go.

Yup, permutations are always a bit nightmarish to work with :cry:

File: src/Data/Fin/Properties.agda (hunk 5 above)
Comment (TOTBWF) [reply src/Data/Fin/Properties.agda:693]: Surprisingly, you do! If you don't, agda gets very confused and yells about not being able to solve some unification problems.

Comment (MatthewDaggitt) [reply src/Data/Fin/Properties.agda:693]: Ah yes because you're matching on `Fin (suc n)` rather than `Fin n`!

**Comment by MatthewDaggitt** [comment]
Thanks for the changes!

**Pull request merged**